### ISSUE TITLE
Files produced in ScanImage 2020.1 cannot be parsed.

### ISSUE THREAD

**Issue by throwaway-scan** [cgohlke/tifffile#46]
We at Vidrio have received a ticket pertaining parsing scanimage tifs with tifffile. Just passing this on:

> In the newest scanimage, metadata from the images has changed and is no longer parseable by https://pypi.org/project/tifffile/ 
> 
> My pipeline relies on this package. It seems the error is when the package tries to check if the metadata is a scanimage image in a not very robust way. I've fixed this for myself but this may affect other users.
> 
>     def is_scanimage(self):
>         """Page contains ScanImage metadata."""
>         return (
>             self.description[:12] == 'state.config'
>             or self.software[:22] == 'SI.LINE_FORMAT_VERSION'
>             or 'scanimage.SI' in self.description[-256:]

**Comment by cgohlke** [comment]
There is not enough information to take action on the issue.

**Comment by HagaiHargil** [comment]
Hi @cgohlke,

I'd be happy if you could re-open this issue. I'm providing [three frames](https://drive.google.com/file/d/1GK_uW1fCQlQMMGZ3RWDMC6mnQWhDzR_5/view?usp=sharing) generated by the latest version of ScanImage which, as the OP correctly notes, aren't identified as a SI file.

I also tried directly using `read_scanimage_metadata()` to no help.

I'd be happy if you could take a look and see for yourself. Thanks.

**Comment by cgohlke** [comment]
Thank you for sharing the file. I'll fix the detection in the next version.

`read_scanimage_metadata()` should work with v2021.2.26 released a couple of days ago:
```Python
from tifffile import read_scanimage_metadata
frame_data, roi_data = read_scanimage_metadata(open('3frames.tif', 'rb'))[:2]
```

A bigger issue is how to reliably get the number of channels, frames, and slices from frame_data for different versions of ScanImage metadata. Any idea?

**Comment by HagaiHargil** [comment]
> read_scanimage_metadata() should work with v2021.2.26 released a couple of days ago:

Ahh, I see. I definitely used the latest and greatest version, but unfortunately my code was identical to yours, with the small change of opening the file in `'r'` mode, instead of `'rb'`. The error I received is:

```
UnicodeDecodeError                        Traceback (most recent call last)
/opt/miniconda3/envs/caiman/lib/python3.7/site-packages/tifffile/tifffile.py in read_scanimage_metadata(fh)
  13314     try:
> 13315         byteorder, version = struct.unpack('<2sH', fh.read(4))
  13316         if byteorder != b'II' or version != 43:

/opt/miniconda3/envs/caiman/lib/python3.7/codecs.py in decode(self, input, final)
    321         data = self.buffer + input
--> 322         (result, consumed) = self._buffer_decode(data, self.errors, final)
    323         # keep undecoded input until the next call

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xd9 in position 28: invalid continuation byte

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<ipython-input-14-9541e1ee7c37> in <module>
      1 from tifffile import read_scanimage_metadata
----> 2 a = read_scanimage_metadata(open('fname.tif', 'r'))

/opt/miniconda3/envs/caiman/lib/python3.7/site-packages/tifffile/tifffile.py in read_scanimage_metadata(fh)
  13323             )
  13324     except Exception as exc:
> 13325         raise ValueError('not a ScanImage BigTIFF v3 or v4 file') from exc
  13326 
  13327     frame_data = matlabstr2py(bytes2str(fh.read(size0)[:-1]))

ValueError: not a ScanImage BigTIFF v3 or v4 file
```
Which led me to believe that there's a bug in that function - I didn't pick up from this message that I should open the file in binary mode, which obviously makes total sense. Could you perhaps see if this error message could be improved?

> A bigger issue is how to reliably get the number of channels, frames, and slices from frame_data for different versions of ScanImage metadata. Any idea?

I'm providing [here]() a V3 file for you to inspect. This file is large - over 3 GB - but since it's from an older SI version I can no longer generate dummy files with a reasonable size for testing purposes, so I'm sorry about that. To get the relevant information from the frame data you can look at the following fields:

Number of channels: `'SI.hChannels.channelSave': [[1], [2]],` - two channels in this TIF, although channels in this case refer to different data sources which are interleaved in the resulting TIF, so tifffile might not care about that. I don't remember what are channels in the TIF spec.
Frames and slices:  `'SI.hStackManager.framesPerSlice': 3000` and  `'SI.hStackManager.numSlices': 1`, which means that this file has 3000 * 1 = 3000 frames (or pages, as it is called I believe in TIFspeak) of data. Even when the number of slices is increased all images are still a part of the same series of data.

Thanks again for your hard work!

**Comment by cgohlke** [comment]
Thank you. Do you know if (in ScanImage lingo) `Slice` refers to depth slice and `Frame` to time frame?

**Comment by throwaway-scan** [comment]
`Slice` refers to the depth when acquiring. Multiple `Frame`s (scanned images) can be acquired at each slice before proceeding to the next slice.

**Comment by cgohlke** [comment]
For the three pages file `3frames.tif` I get `SI.hStackManager.numSlices = 150` and `SI.hChannels.channelSave` is an `int`, not a `list`. There's an `SI.hStackManager.actualNumSlices` that seems correct.

**Comment by HagaiHargil** [comment]
Regarding `numSlices` - I wasn't aware that there's an `actualNumSlices` as well, which isn't present in the v3 file. It seems to refer to two slightly different settings in the software, but the naming isn't too great. 
Regarding `channelSave` - I went looking for a v4 file with two channels interleaved and (sadly?) the format is still `[[1], [2]]` in that case, so it seems you have to prepare for either a single number or for a list of lists of numbers.

Do you need these parameters to preallocate correctly some buffers for `imread`? Or do they have some other use?

**Comment by cgohlke** [comment]
> Regarding numSlices - I wasn't aware that there's an actualNumSlices as well, which isn't present in the v3 file. It seems to refer to two slightly different settings in the software, but the naming isn't too great.

If `numSlices` is not reliable, and `actualNumSlices` is a different setting, what is the actual real number of slices found in a ScanImage file? Should it be calculated from the number of pages, `channelSave`, and `framesPerSlice` (which might be `inf` btw)?

> Do you need these parameters to preallocate correctly some buffers for imread? Or do they have some other use?

Those parameters are needed to shape the array returned by imread: `(slices, frames, channels, length, width)`.  See [Image File Storage & Format (r3.8)](http://scanimage.vidriotechnologies.com/pages/viewpage.action?pageId=17400312). But maybe that storage format does not apply to newer versions?

**Comment by HagaiHargil** [comment]
The number of pages is always correct obviously, and its value is what you're looking for, but I'm guessing it's not known in advance? I.e. you have to iterate over the IFDs to know that? 

SI TIFs are still packed as that document describes, but the all the dimensions you'll need are the number of pages (dimension 0), length and width. If you can't use the number of pages then it seems you need to calculate `numSlices` * `framesPerSlice` * `len(channelSave)`, and replace `numSlices` with `actualNumSlices` if it exists. Not sure when will `framesPerSlice` be `inf`, and there might be other edge cases with frame averaging, although I'm not sure how it will come into play here exactly.

**Comment by cgohlke** [comment]
Should be fixed in v2021.3.4.
